### PR TITLE
[feat] add flash attention

### PR DESCRIPTION
**Pull request by oahzxl**
add cuda and triton flash attention

### CODE REVIEW

**Comment by oahzxl** [comment]
image

File: tests/test_utils/test_flash_attention.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    # triton implementation
+    q, k, v = map(lambda x: rearrange(x, 'z h n d -> (z n) h d'), [q, k, v])
+    tri_out = flash_attention(q, k, v, sm_scale, Z, N_CTX)
+    dout = rearrange(dout, 'z h n d -> (z n) h d').detach()
+    tri_out.backward(dout, retain_graph=True)
+    tri_dq, tri_dk, tri_dv, = torch.autograd.grad(tri_out, (q, k, v), dout)
+    tri_out, tri_dq, tri_dk, tri_dv = map(lambda x: rearrange(x, '(z n) h d -> z h n d', z=Z), (tri_out, tri_dq, tri_dk, tri_dv))
+
+    # compare
+    assert 1e-4 > torch.mean(torch.abs(ref_out - tri_out))
```
Comment (feifeibear) [inline tests/test_utils/test_flash_attention.py:83]: torch.allclose

File: colossalai/kernel/cuda_native/flash_attention.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+This is a Triton implementation of the Flash Attention algorithm
+(see: Dao et al., https://arxiv.org/pdf/2205.14135v2.pdf; Rabe and Staats https://arxiv.org/pdf/2112.05682v2.pdf; Triton https://github.com/openai/triton)
+"""
+
+import torch
+from subprocess import PIPE, run
+
+import triton
+import triton.language as tl
+
```
Comment (feifeibear) [inline colossalai/kernel/cuda_native/flash_attention.py:13]: try:
     from flash_attn.flash_attn_interface import flash_attn_unpadded_func
catch ImportError:
     print('please install flash_attn, from url xxxx')

File: tests/test_utils/test_flash_attention.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
+import torch
+import pytest
+from einops import rearrange
+from colossalai.kernel.cuda_native.flash_attention import flash_attention, triton_flash_attention, triton_check
+
+
+TRITON_AVALIABLE = triton_check()
```
Comment (feifeibear) [inline tests/test_utils/test_flash_attention.py:7]: from colossalai.kernel.cuda_native.flash_attention import TRITON_AVALIABLE
do not need to run it again?

File: tests/test_utils/test_flash_attention.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    else:
+        try:
+            tri_out = flash_attention(q, k, v, sm_scale, Z, N_CTX)
+        except RuntimeError:
+            pass
+        else:
+            raise TypeError("Error type not match!")
+
+
+@pytest.mark.parametrize('Z, H, N_CTX, D_HEAD', [(3, 2, 2048, 64)])
```
Comment (feifeibear) [inline tests/test_utils/test_flash_attention.py:52]: shrink N_CTX and D_HEAD to accelerate the unitest.

File: tests/test_utils/test_flash_attention.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+def test_flash_attention(Z, H, N_CTX, D_HEAD, dtype=torch.float16):
+    torch.manual_seed(20)
+    q = torch.randn((Z, H, N_CTX, D_HEAD), dtype=dtype, device="cuda").normal_(mean=0, std=.5).requires_grad_()
+    k = torch.randn((Z, H, N_CTX, D_HEAD), dtype=dtype, device="cuda").normal_(mean=0, std=.5).requires_grad_()
+    v = torch.randn((Z, H, N_CTX, D_HEAD), dtype=dtype, device="cuda").normal_(mean=0, std=.5).requires_grad_()
+    sm_scale = 0.3
+    dout = torch.randn_like(q)
+    
+    # reference implementation
```
Comment (feifeibear) [inline tests/test_utils/test_flash_attention.py:61]: Can you wrap the attention reference implementation into a function?

File: colossalai/kernel/cuda_native/flash_attention.py [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+        return False
+    return False
+
+TRITON_AVALIABLE = triton_check()
+
+
+@triton.jit
+def _fwd_kernel(
+    Q, K, V, sm_scale,
+    TMP, L, M,  # NOTE: TMP is a scratchpad buffer to workaround a compiler bug
```
Comment (feifeibear) [inline colossalai/kernel/cuda_native/flash_attention.py:37]: can you add type hints for each parameter?
for example
TMP : torch.Tensor, L : int

Comment (oahzxl) [reply colossalai/kernel/cuda_native/flash_attention.py:43]: Not sure about some param. Will update it with alphafold version flash attention

Comment (feifeibear) [reply colossalai/kernel/cuda_native/flash_attention.py:43]: OK

**Review by feifeibear — APPROVED** [review]

**Pull request merged**